app/javascript/utils/react-bootloader.tsx

```diff
@@ -0,0 +1,181 @@
+
+if (process.env.BUGSNAG_API_KEY) {
+  Bugsnag.start({
+    apiKey: process.env.BUGSNAG_API_KEY,
+    releaseStage: process.env.NODE_ENV,
+    plugins: [new BugsnagPluginReact()],
+    enabledReleaseStages: ['production'],
+    collectUserIp: false,
+    onError: function (event) {
+      const tag = document.querySelector<HTMLMetaElement>(
+        'meta[name="user-id"]'
+      )
+
+      if (!tag) {
+        return true
+      }
+
+      event.setUser(tag.content)
+    },
+  })
+  const reactPlugin = Bugsnag.getPlugin('react')
+  if (reactPlugin) {
+    ErrorBoundary = reactPlugin.createErrorBoundary(React)
+  } else {
+    throw new Error('Bugsnag React plugin is null')
+  }
+}
+
+function initEventListeners() {
+  // This changes any extra things that need changing from the
+  // turbo frame, such as body class or page title
+  document.addEventListener('turbo:frame-render', (e) => {
+    const event = e as CustomEvent<TurboFrameRenderDetail>
+    const bodyClass = event.detail.fetchResponse.response.headers.get(
+      'exercism-body-class'
+    )
+    if (bodyClass) {
+      document.body.className = bodyClass
+    }
+  })
+
+  document.addEventListener('turbo:before-fetch-request', () => {
+    setTurboStyle('* { cursor:wait !important }')
+  })
+
+  document.addEventListener('turbo:before-render', () => {
+    window.scrollTo({
+      top: 0,
+      left: 0,
+      behavior: 'auto',
+    })
+    setTurboStyle('')
+  })
+}
+
+/********************************************************/
+/** Add a loading cursor when a turbo-frame is loading **/
+/********************************************************/
+const styleElemId = 'turbo-style'
+const setTurboStyle = (style: string) => {
+  let styleElem = document.getElementById(styleElemId)
+  if (!styleElem) {
+    styleElem = document.createElement('style')
+    styleElem.id = styleElemId
+    document.head.appendChild(styleElem)
+  }
+  styleElem.textContent = style
+}
+
+export function initReact(mappings: Mappings): void {
+  const renderThings = (parentElement = document.body) => {
+    renderComponents(parentElement, mappings)
+    renderTooltips(parentElement, mappings)
+  }
+
+  // This adds rendering for all future turbo clicks
+  document.addEventListener('turbo:load', () => {
+    renderThings()
+  })
+
+  // This renders if turbo has already finished at the
+  // point at which this calls. See packs/core.tsx
+  if (window.turboLoaded) {
+    renderThings()
+  }
+}
+
+const render = (elem: HTMLElement, component: React.ReactNode) => {
+  ReactDOM.render(
+    <React.StrictMode>
+      <ReactQueryCacheProvider queryCache={window.queryCache}>
+        <ErrorBoundary>{component}</ErrorBoundary>
+      </ReactQueryCacheProvider>
+    </React.StrictMode>,
+    elem
+  )
+  document.addEventListener('turbo:before-frame-render', () => {
+    ReactDOM.unmountComponentAtNode(elem)
+  })
+}
+
+export function renderComponents(
+  parentElement: HTMLElement = document.body,
```

Comment:
changed all
```js
 if (!parentElement) {
    parentElement = document.body
  }
  ```
  into default parameters

Reply:
Isn't that changing the behavior though? Previously, these calls would end up with `parentElement` being set to `document.body`, but they won't anymore:

```js
renderComponents(null, ...)
renderComponents(undefined, ...)
```

Reply:
@dem4ron Did you see this comment?

Reply:
In our case this doesn't change the behaviour, because we don't pass in `null`s
I made sure things wouldn't fail, by passing in a default parameter a level above too, which used to be simply `undefined`
```diff
- const renderThings = (parentElement) => {
+ const renderThings = (parentElement = document.body) => {
    renderComponents(parentElement, mappings)
    renderTooltips(parentElement, mappings)
  }

  // This adds rendering for all future turbo clicks
  document.addEventListener('turbo:load', () => {
    renderThings()
  })
  ...
```


Another place where `renderComponents` is used is in `LazyTippy.tsx`:

```ts
export const LazyTippy = (props: LazyTippyProps): JSX.Element => {
  const { renderReactComponents, ...tippyProps } = props
  const [mounted, setMounted] = React.useState(false)

  const lazyPlugin = {
    fn: () => ({
      onMount: () => setMounted(true),
      onHidden: () => setMounted(false),
      onAfterUpdate: (instance: Instance) => {
        if (renderReactComponents) {
          renderComponents(instance.popper, mappings)
        }
      },
    }),
  }
  ```
  
In the worst-case scenario here, the value would be `undefined` too
  
On the other hand - given how mission critical this code is - I'm leaning towards reverting these changes to the input validation approach, because that is significantly more robust.

Reply:
But here:
```ts
  const renderThings = (parentElement: HTMLElement) => {
    console.log(parentElement)
    if (!parentElement) {
      parentElement = document.body
    }
    renderComponents(parentElement, mappings)
    renderTooltips(parentElement, mappings)
  }
  ```
  
  I'd just pass in `document.body`